servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,661 @@
+                }
+            }
+        }
+
+        private void distributeMappedDemand(FlatMapPublisherSubscriber<T, R> hungrySubscriber) {
+            final int quota = reserveMappedDemandQuota();
+            if (quota > 0) {
+                final int usedQuota = hungrySubscriber.request(quota);
+                if (usedQuota < quota && mappedDemandUpdater.getAndAccumulate(this, quota - usedQuota,
+                        FlowControlUtils::addWithOverflowProtection) == 0) {
+                    // If we gave some back and transitioned from 0 demand, we need to try to distribute demand
+                    // in case other hungry subscribers were added in the mean time
+                    // (since we have not acquired the requestingLock).
+                    distributeMappedDemand();
+                }
+            } else { // slow path. no demand, add to queue and process later when demand arrives.
+                hungrySubscribers.add(hungrySubscriber);
+                distributeMappedDemand();
+            }
+        }
+
+        private void distributeMappedDemand() {
+            Throwable delayedCause = null;
+            boolean tryAcquire = true;
+            while (tryAcquire && tryAcquireLock(requestingLockUpdater, this)) {
+                try {
+                    final long availableRequestN = reserveMappedDemand();
+                    if (availableRequestN > 0) {
+                        long remainingRequestN = availableRequestN;
+                        final int quota = calculateRequestNQuota(availableRequestN);
+                        FlatMapPublisherSubscriber<T, R> hungrySubscriber;
+                        while ((hungrySubscriber = hungrySubscribers.poll()) != null) {
+                            remainingRequestN -= hungrySubscriber.request(quota);
+                        }
+                        if (remainingRequestN > 0) {
+                            mappedDemandUpdater.accumulateAndGet(this, remainingRequestN,
+                                    FlowControlUtils::addWithOverflowProtection);
+                        }
+                    }
+                } catch (Throwable cause) {
+                    delayedCause = catchUnexpected(delayedCause, cause);
+                } finally {
+                    tryAcquire = !releaseLock(requestingLockUpdater, this);
+                }
+            }
+            if (delayedCause != null) {
+                throwException(delayedCause);
+            }
+        }
+
+        private int calculateRequestNQuota(long availableRequestN) {
+            // Get an approximate quota to distribute to each active mapped subscriber. If
+            // activeMappedSources changes (subscriber added or terminated) we will re-distribute.
+            final int prevActiveSources = activeMappedSources;
+            return (int) min(Integer.MAX_VALUE,
+                    max(prevActiveSources > 0 ? availableRequestN / prevActiveSources : availableRequestN,
```

Comment:
the `mappedDemand` accounting is not quite right and needs some fixing. the intention was to at least distribute `minMappedDemand` to each source so that each source is allowed to produce data while capping the queue size at `minMappedDemand * maxConcurrency`. let me re-work this a bit.

using `availableRequestN` was an optimization geared toward steady state to get a more accurate estimate which should result in larger demand chunks being distributed and less re-distribution events. It would be unfair during non-steady state as you indicated. This optimization is not critical and I will remove it for now.